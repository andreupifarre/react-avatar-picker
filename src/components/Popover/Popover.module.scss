.Popover {
  width: 280px;
  box-sizing: content-box;
  background-color: rgb(44, 48, 51);
  transform: translateX(-109px) translateY(10px);
  border-radius: 2px;
  box-shadow: 2px 2px 10px 0px rgb(102, 102, 102);
  padding: 0 0 15px;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.Popover::after {
  content: '';
  position: absolute;
  top: -16px;
  left: calc(50% - 8px);
  border-top: 8px solid transparent;
  border-bottom: 8px solid rgb(44, 48, 51);
  border-left: 8px solid transparent;
  border-right: 8px solid transparent;
}

.Title {
  color: rgba(249, 249, 249, 0.9);
  font-size: 16px;
  margin: 14px 0 12px;
}

.AvatarImagesList {
  margin: 0;
  padding: 0;
  list-style: none;
  width: 260px;
  display: grid;
  gap: 6px 4px;
  grid-template-areas: '. . . .';
}

.AvatarImageItem {
  position: relative;
  border-radius: 50%;
  width: 54px;
  height: 54px;
  border: 3px solid transparent;
}

.AvatarImageItemSelected {
  border-color: rgb(122, 161, 178);
}

.AvatarImageItem:hover {
  border-color: rgb(155, 160, 163);
}

.AvatarImageSpinner,
.AvatarImageSpinner:after {
  border-radius: 50%;
  width: 54px;
  height: 54px;
}
.AvatarImageSpinner {
  font-size: 10px;
  position: absolute;
  top: -3px;
  left: -3px;
  text-indent: -9999em;
  border-top: 3px solid rgb(122, 161, 178);
  border-right: 3px solid rgb(122, 161, 178);
  border-bottom: 3px solid rgb(122, 161, 178);
  border-left: 3px solid rgb(44, 48, 51);
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
  -webkit-animation: LoadingSpinnerAnimation 900ms infinite linear;
  animation: LoadingSpinnerAnimation 900ms infinite linear;
}

@-webkit-keyframes LoadingSpinnerAnimation {
  0% {
    -webkit-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes LoadingSpinnerAnimation {
  0% {
    -webkit-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

.AvatarImageOverlay {
  pointer-events: none;
  position: absolute;
  width: inherit;
  height: inherit;
  border-radius: inherit;
  background-color: rgb(122, 161, 178);
  opacity: 0;
  top: 0;
}

.AvatarImageItem:hover .AvatarImageOverlay {
  opacity: 0.2;
}
