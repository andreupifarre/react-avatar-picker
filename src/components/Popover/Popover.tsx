import React from 'react';
import classNames from 'classnames';
import { AvatarImageItemType } from '../AvatarImage/AvatarImageItemType';
import AvatarImage from '../AvatarImage/AvatarImage';

import styles from './Popover.module.scss';

type AvatarImageProps = {
  items: AvatarImageItemType[];
  selected: number;
  loading?: number;
  onSelected: (id: number) => void;
};

const Popover = ({
  items,
  onSelected,
  selected,
  loading,
}: AvatarImageProps) => {
  const handleClick = (id: number) => {
    onSelected(id);
  };

  return (
    <div className={styles.Popover}>
      <h2 className={styles.Title}>Choose your avatar</h2>
      <ul className={styles.AvatarImagesList}>
        {items.map((item, i: number) => (
          <li
            className={classNames(styles.AvatarImageItem, {
              [styles.AvatarImageItemSelected]: selected === item.id,
            })}
            key={i}
          >
            <AvatarImage
              id={item.id}
              src={item.src}
              label={item.label}
              onClick={handleClick}
              size={54}
            />
            {loading === item.id && (
              <div className={styles.AvatarImageSpinner} />
            )}
            <div className={styles.AvatarImageOverlay} />
          </li>
        ))}
      </ul>
    </div>
  );
};

export default Popover;
