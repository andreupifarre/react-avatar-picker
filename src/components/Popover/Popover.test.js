import React from 'react';
import { shallow } from 'enzyme';

import Popover from './Popover';

const render = (props) => shallow(<Popover {...props} />);

const items = [
  { src: 'avatar1.png', label: 'Avatar 1', id: 1 },
  { src: 'avatar2.png', label: 'Avatar 2', id: 2 },
  { src: 'avatar3.png', label: 'Avatar 3', id: 3 },
  { src: 'avatar4.png', label: 'Avatar 4', id: 4 },
  { src: 'avatar5.png', label: 'Avatar 5', id: 5 },
  { src: 'avatar6.png', label: 'Avatar 6', id: 6 },
];
const selected = 1;
const loading = 2;

describe('Popover', () => {
  it('should display the same amount of AvatarImage(s) as objects passed to items array', () => {
    const wrapper = render({ items, selected, loading });

    expect(wrapper.find('li').length).toEqual(6);
  });

  it('should have the first li selected with the class .AvatarImageItemSelected', () => {
    const wrapper = render({ items, selected, loading });

    expect(
      wrapper.find('.AvatarImageItemSelected').find('AvatarImage').prop('id'),
    ).toEqual(1);
  });

  it('should render correctly', () => {
    expect(
      render({ items, selected, loading }).getElements(),
    ).toMatchSnapshot();
  });
});
