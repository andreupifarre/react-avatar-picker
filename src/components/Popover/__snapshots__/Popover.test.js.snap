// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should render correctly 1`] = `
Array [
  <div
    className="Popover"
  >
    <h2
      className="Title"
    >
      Choose your avatar
    </h2>
    <ul
      className="AvatarImagesList"
    >
      <li
        className="AvatarImageItem AvatarImageItemSelected"
      >
        <AvatarImage
          id={1}
          label="Avatar 1"
          onClick={[Function]}
          size={54}
          src="avatar1.png"
        />
        <div
          className="AvatarImageOverlay"
        />
      </li>
      <li
        className="AvatarImageItem"
      >
        <AvatarImage
          id={2}
          label="Avatar 2"
          onClick={[Function]}
          size={54}
          src="avatar2.png"
        />
        <div
          className="AvatarImageSpinner"
        />
        <div
          className="AvatarImageOverlay"
        />
      </li>
      <li
        className="AvatarImageItem"
      >
        <AvatarImage
          id={3}
          label="Avatar 3"
          onClick={[Function]}
          size={54}
          src="avatar3.png"
        />
        <div
          className="AvatarImageOverlay"
        />
      </li>
      <li
        className="AvatarImageItem"
      >
        <AvatarImage
          id={4}
          label="Avatar 4"
          onClick={[Function]}
          size={54}
          src="avatar4.png"
        />
        <div
          className="AvatarImageOverlay"
        />
      </li>
      <li
        className="AvatarImageItem"
      >
        <AvatarImage
          id={5}
          label="Avatar 5"
          onClick={[Function]}
          size={54}
          src="avatar5.png"
        />
        <div
          className="AvatarImageOverlay"
        />
      </li>
      <li
        className="AvatarImageItem"
      >
        <AvatarImage
          id={6}
          label="Avatar 6"
          onClick={[Function]}
          size={54}
          src="avatar6.png"
        />
        <div
          className="AvatarImageOverlay"
        />
      </li>
    </ul>
  </div>,
]
`;
